### ISSUE TITLE
CS8629 triggered after certain "is object" or "is null" Debug assertions

### ISSUE THREAD

**Issue by nathan-alden-sr** [dotnet/roslyn#39540]
**Version Used**: 

.NET Core 3.0.100
Visual Studio 2019 16.3.5

**Steps to Reproduce**:

The following methods produce CS8629 warnings seemingly incorrectly:
```csharp
using System.Diagnostics;

#nullable enable

namespace A
{
    public class B
    {
        public void Foo(int? foo)
        {
            Debug.Assert(foo is object); // Should guarantee that foo is not null
            // This actually prevents the warning:
            // Debug.Assert(!(foo is null));

            int _ = foo.Value; // CS8629
        }

        public void Bar(object? bar)
        {
            Debug.Assert(bar is object); // Should guarantee that bar is not null
            Debug.Assert(!(bar is null)); // Should guarantee that bar is not null

            string _ = bar.ToString(); // CS8629
        }
    }
}
```
Project file:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.0</TargetFramework>
  </PropertyGroup>
</Project>
```
**Expected Behavior**:

No CS8629 warnings should be generated.

**Actual Behavior**:

Two CS8629 warnings are generated.

**Comment by nathan-alden-sr** [comment]
I admit I'm not an expert in what `is object` *should* do, but IMO these warnings violate the principle of least astonishment.

**Comment by sharwell** [comment]
:memo: The first one (method `Foo`) is a duplicate of #38213.

**Comment by jcouv** [comment]
Verified this was fixed in 16.5p1 or earlier
